vowpalwabbit/parse_args.cc

```diff
@@ -1272,98 +1275,132 @@ VW::LEARNER::base_learner* setup_base(options_i& options, vw& all)
 {
   auto setup_func = all.reduction_stack.top();
   all.reduction_stack.pop();
-  auto base = setup_func(options, all);
+  auto base = std::get<1>(setup_func)(options, all);
 
-  if (base == nullptr)
-    return setup_base(options, all);
+  // returning nullptr means that setup_func (any reduction) was not 'enabled' but
+  // only added their respective command args and did not add itself into the
+  // chain of learners, therefore we call into setup_base again
+  if (base == nullptr) { return setup_base(options, all); }
   else
+  {
+    all.enabled_reductions.push_back(std::get<0>(setup_func));
     return base;
+  }
+}
+
+void register_reductions(vw& all, std::vector<VW::LEARNER::base_learner* (*)(VW::config::options_i&, vw&)>& reductions)
+{
+  std::map<VW::LEARNER::base_learner* (*)(VW::config::options_i&, vw&), std::string> allowlist = {{GD::setup, "gd"},
```

Comment:
A typedef for the function pointer would make this code easier to read.  Can also be reused in the method signature above.

Reply:
good idea

Reply:
done